### ISSUE TITLE
[Bug] Possible Block/Unblock apps issue

### ISSUE THREAD

**Issue by ryanedlund** [PSAppDeployToolkit/PSAppDeployToolkit#1217]
### Prerequisites

- [X] Ensure you write a short, descriptive title after [Bug] above.
- [X] Make sure to [search for any existing issues](https://github.com/psappdeploytoolkit/psappdeploytoolkit/issues) before filing a new one.
- [X] Verify you are able to reproduce the issue with the [latest released version](https://www.github.com/psappdeploytoolkit/psappdeploytoolkit/releases/latest)

### PSAppDeployToolkit version

4.0.4

### Describe the bug

Not sure if it's just because I'm doing something wrong with my script (due to new v4 syntax or something  - running 4.0.4) or if this is a bug, but now that apps are actually getting blocked (Issue #1216) - although I'm not getting any popup, they just won't launch - I now seem to be having an issue with blocked apps not unblocking properly as part of the normal close-ADTsession process. I see the log stating that the registry key to block execution is being populated, but not the one removing it afterward, although I do see the scheduled task is removed properly (and is logged as such).

Looking at the module, it appears the  Unblock-ADTAppExecutionInternal function that does the actual reg key removal is looking for the "Debugger" registry value to contain "Show-ADTBlockedAppDialog" in order for it to remove the value, however, looking at the contents of my "Debugger" value I see the following:

conhost.exe --headless powershell.exe -NonInteractive -NoProfile -NoLogo -Command & ([scriptblock]::Create([Microsoft.Win32.Registry]::GetValue('HKEY_LOCAL_MACHINE\SOFTWARE\PSAppDeployToolkit', 'BlockExecutionCommand', $null)))

That matches what I'm seeing in the log as well:

[Pre-Install] :: Setting registry key value: [Microsoft.PowerShell.Core\Registry::HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows NT\CurrentVersion\Image File Execution Options\ms-teams.exe] [Debugger = conhost.exe --headless powershell.exe -NonInteractive -NoProfile -NoLogo -Command & ([scriptblock]::Create([Microsoft.Win32.Registry]::GetValue('HKEY_LOCAL_MACHINE\SOFTWARE\PSAppDeployToolkit', 'BlockExecutionCommand', $null)))].	Set-ADTRegistryKey	12/24/2024 2:08:16 PM	15904 (0x3E20)

It seems like either the block action isn't setting the Debugger reg key as expected, or the unblock action is looking for something that doesn't exist.

Assuming that what the Debugger should contain is actually the value stored in the "BlockExecutionCommand" value in HKEY_LOCAL_MACHINE\SOFTWARE\PSAppDeployToolkit, even that doesn't appear to contain the expected "Show-ADTBlockedAppDialog" needed in order to remove the item:

& (Import-Module -FullyQualifiedName @{ ModuleName = 'c:\apps\Package - v4\PSAppDeployToolkit\PSAppDeployToolkit.psd1'; Guid = '8c3c366b-8606-4576-9f2d-4051144f7ca2'; ModuleVersion = '4.0.4' } -PassThru) { & $CommandTable.'Initialize-ADTModule' -ScriptDirectory 'c:\apps\Package - v4'; $null = & $CommandTable.'Show-ADTInstallationPromptFluent' -Title 'Vendor App 1.0.0.0' -Subtitle 'PSAppDeployToolkit - App Install' -Timeout 3300 -Message 'Launching this application has been temporarily blocked so that an installation operation can complete.' -Icon Warning -ButtonRightText OK }

As a side note, should there be something in the unblock process that removes the "BlockExecutionCommand" value from HKEY_LOCAL_MACHINE\SOFTWARE\PSAppDeployToolkit, as part of the cleanup, because I'm not finding that if so?

### Steps to reproduce

1. In app package, I have set:  Show-ADTInstallationWelcome -CloseProcesses ms-teams -ForceCloseProcessesCountdown 3600 -PersistPrompt -BlockExecution
2. This results in the block registry keys and scheduled task being added to the processes indicated.  When the close-ADTsession process runs those keys and tasks should be getting cleaned up, but the unblock action that should be removing the reg key(s) is not occurring.

[2024-12-24 14:44:13.658] [Finalization] [Unblock-ADTAppExecution] [Info] :: Deleting Scheduled Tasks ['PSAppDeployToolkit_Vendor_App_1.0.0.0_x64_EN_01_BlockedApps'].
[2024-12-24 14:44:13.673] [Finalization] [Close-ADTInstallationProgressFluent] [Info] :: Closing the installation progress dialog.
[2024-12-24 14:44:13.739] [Finalization] [Close-ADTSession] [Success] :: [Vendor_App_1.0.0.0_x64_EN_01] install completed with exit code [0].
[2024-12-24 14:44:13.742] [Finalization] [Close-ADTSession] [Info] :: *******************************************************************************
VERBOSE: Removing the imported "Write-ADTLogEntry" function.

### Environment data

```console
OsName               : Microsoft Windows 10 Enterprise
OsOperatingSystemSKU : EnterpriseEdition
OsArchitecture       : 64-bit
WindowsVersion       : 2009
WindowsProductName   : Windows 10 Enterprise
WindowsBuildLabEx    : 19041.1.amd64fre.vb_release.191206-1406
OsLanguage           : en-US
OsMuiLanguages       : {en-US}
KeyboardLayout       : en-US
TimeZone             : (UTC-05:00) Eastern Time (US & Canada)
HyperVisorPresent    : True
CsPartOfDomain       : True
CsPCSystemType       : Desktop

.NET SDK (reflecting any global.json):
 Version:   6.0.428
 Commit:    ef6f5ce48c

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.19045
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\6.0.428\

Host:
  Version:      7.0.20
  Architecture: x64
  Commit:       0fb6ac59fb

.NET SDKs installed:
  6.0.428 [C:\Program Files\dotnet\sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.App 6.0.36 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 6.0.36 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 7.0.20 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 6.0.36 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 7.0.20 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]

Other architectures found:
  None

Environment variables:
  Not set

global.json file:
  Not found
```

**Comment by mjr4077au** [comment]
This has already been identified and resolved in https://github.com/PSAppDeployToolkit/PSAppDeployToolkit/commit/e05a91b48b34b01137f526119b50536e329fc9f4, it's just not in a release yet. Can you download the zip file from this action and test? It's a candidate build of 4.0.5: https://github.com/PSAppDeployToolkit/PSAppDeployToolkit/actions/runs/12453354484

**Comment by ryanedlund** [comment]
That updated code does appear to have addressed the clean-up of debugger values under HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows NT\CurrentVersion\Image File Execution Option.

The "BlockExecutionCommand" value is still under HKEY_LOCAL_MACHINE\SOFTWARE\PSAppDeployToolkit, but it looks like the data for that value for that is just overwritten by the next package to use -BlockExecution.


Is the data that is being written to that Debugger value what we are expecting to see? The value seems odd to me, but not sure what "good" looks ***@***.***


What makes me question the validity of that is the fact that, with it in place as-is, I don't get any kind of pop-up stating that the app is blocked due to running PSADT session when trying to launch the specific app, and I'm pretty sure that's what that is supposed to do...

If I manually run the contents of the "BlockExecutionCommand" value in Powershell I do see the expected prompt pop-up.


Ryan Edlund | OS Platform Engineer III | The Huntington National Bank | Phone: 330-996-6307 | Mobile: 234-206-1007 | huntington.com<https://www.huntington.com/>

***@***.***


Internal Use


Internal Use

**Comment by mjr4077au** [comment]
The value for BlockExecutionCommand doesn't get cleaned up for the reasons you've just mentioned, because it gets overwritten on the next execution anyway. 

The value is what we're expecting as you can't have an IFEO value with a command path greater than 255 characters, and with the module-based setup we were exceeding this very easily. The only other way around it would be if we shipped a script with the module specifically for this purpose to contain the required logic, or lose things like the app-specific title etc. However it can still fall short depending on how long the path to the script is, etc. 

If it's not executing properly, my recommended goto here would be to enable system-wide PowerShell transcription, which can either be done via a configuration profile or `gpedit.msc` locally (refer to https://admx.help/?Category=Windows_10_2016&Policy=Microsoft.Policies.PowerShell::EnableTranscripting).

Once this is enabled to your specified directory, any error output from PowerShell will be captured to text file and will give you proper visibility into what the go is. If nothing is captured, then PowerShell.exe isn't running at all, and that's still helpful to know here.

**Comment by ryanedlund** [comment]
Enabled global transcription, but do not see any logs when attempting to launch a blocked ms-teams.exe application.  I suspect that may be due to the fact that Teams is now a UWP app, but not entirely sure.

Confirmed I do see other scripts creating logs - such as the PSADT invoke-AppDeployToolkit.exe itself when attempting a re-install - in the defined location, so I know that is working, just not seeing anything when debugger key is active and ms-teams app launch is attempted.

Changed to another regular win32 exe (KeePass.exe), and found I am now getting a transcription log created for that, which I have attached.

Did some playing with the IFEO Debugger value, and I found that the behavior seen the log seems to be the norm: it will pass the full path of the process being debugged as a parameter to the defined "debugger" as defined in that reg key.  For example, using the same "KeePass.exe" IFEO item as we have in the attached log, if "C:\windows\notepad.exe" is defined for the debugger value (instead of the conhost.exe....), notepad will open, but with the exe as the file it is trying to open, as it passed the path of the file to the debugger.



Ryan Edlund | OS Platform Engineer III | The Huntington National Bank | Phone: 330-996-6307 | Mobile: 234-206-1007 | huntington.com<https://www.huntington.com/>

***@***.***


Internal Use


Internal Use

**Comment by mjr4077au** [comment]
UWP apps unfortunately cannot be blocked in this manner. This is known about and is an open issue here: https://github.com/PSAppDeployToolkit/PSAppDeployToolkit/issues/309. We're still surveying options but it doesn't look good. 

The log you've given me is great and shows the issue completely. I'll look at amending that issue with the path the OS passes through to powershell.exe. In the IFEO debugger value, you could probably put `; #` at the end as a test. I'm sure it'll fix the problem.

**Comment by ryanedlund** [comment]
Reading over that issue, my issue with Teams may be related, but not really the same thing, as that seems like that poster was having an issue when referencing UWP apps by their UWP app name instead of the process. I have no problem detecting, closing, and blocking Teams (using ms-teams) from launching with the current workflow, as the process for that is ms-teams.exe. I just don't get the Blocked App popup when attempting to re-launch, it just does nothing.

My guess is the poster for #309 would not have an issue / be in the same boat as me if they used 'PaintStudio.View' (for Paint 3D, aka the MSPaint UWP app in Win10), or 'mspaint' (for Paint in Win10/Win11 - mspaint is UWP in Win11), and 'PhotosApp' for the Photos app instead of their UWP app names.  I am a bit curious where they got the syntax for using UWP app names for that in the first place, and I'm especially curious where the "=Photos" or "=Paint" at the end of the defined name came from, as that's not part of the normal UWP app name. It would be interesting to note if, when they say the app was blocked for Paint, if it actually triggered the blocked app popup, or just prevented execution.


I just confirmed that adding the " #" to the end of the Debugger value data did get the Blocked App popup to fire for a regular win32 app - still had no effect for Teams.


Ryan Edlund | OS Platform Engineer III | The Huntington National Bank | Phone: 330-996-6307 | Mobile: 234-206-1007 | huntington.com<https://www.huntington.com/>

***@***.***


Internal Use


Internal Use

**Comment by mjr4077au** [comment]
Thank you for working with me on this, and for your thorough replies. I'll get the fix in that you confirmed worked when I can, and I'll specifically look at teams to see what we can do there as your feedback is interesting and implies it should work even being a UWP app. I'm interstate on holidays so I wouldn't expect much until just after new years, but I'll get onto this when I return for sure.

**Comment by mjr4077au** [comment]
I've added the `; #` stuff in https://github.com/PSAppDeployToolkit/PSAppDeployToolkit/commit/a8682277f3234ac480f5c0b91bfab213eed3fba5, but I can't look at the ms-teams issue specifically as on my device, it's not blocking as expected for a UWP application:
![image](https://github.com/user-attachments/assets/11cd4f01-b714-449f-92d4-63578fcbf612)
Can you please re-test on your side and confirm ms-teams.exe is properly closed before attempting to open it? It may have had some kind of background running process, or it might just be a UWP application doing as it's well known for.

**Comment by ryanedlund** [comment]
I had confirmed this in all my testing.

Being a UWP app, the start menu shortcut being used to launch the app may not be launching the exe directly - might be a service (like StartMenuExperienceHost) launching the exe. I can attempt to run the actual EXE directly and see what happens, but that's not how an end-user would typically launch the thing.

I assume I'll need to wait for an update to the build in order for the a868227<https://github.com/PSAppDeployToolkit/PSAppDeployToolkit/commit/a8682277f3234ac480f5c0b91bfab213eed3fba5> change to be usable in the standard release? Not sure how I would make use of that as-is as that appears to be a ps1 file used in generating the contents of the module.



One other unrelated note/question - I can submit a feature request or something if needed:  Would it be possible to allow the "Frontend" folder to be removed from the PSAppDeployToolkit directory? It seems like a waste of disk space to have that in there, when the frontend isn't actually used from there - and only v3 or v4 would be, not both. Right now when I attempt to remove that folder it breaks the package saying the toolkit folder has been modified. TBH I would also like to remove "Assets", "Config" and possibly "Strings" as well seeing as how those are not actually used from within that directory either, and end up just being duplicate files.


Ryan Edlund | OS Platform Engineer III | The Huntington National Bank | Phone: 330-996-6307 | Mobile: 234-206-1007 | huntington.com<https://www.huntington.com/>

***@***.***


Internal Use


Internal Use

**Comment by mjr4077au** [comment]
You can download a test build from the latest 4.0.x artifact here: https://github.com/PSAppDeployToolkit/PSAppDeployToolkit/actions/runs/12584028466. 4.0.5 will be out some time this month where it will be included. 

Regarding the UWP findings, that's mostly correct though I couldn't get the block to work even when running ms-teams.exe direct. It's just not a supported app type with IFEO.

Regarding the feature request, we won't be able to help with that as the module needs to be self-contained. While right now we provide a v3 and v4 template, those are generated using that constant via `New-ADTTemplate`, and we'll expect people to generate their own in the future. The module also loads the default config and strings before importing your custom versions, so they're mandatory and cannot be deleted either.

**Comment by mjr4077au** [comment]
Regarding this one, I'm going to close it off for now as I believe the main underlying issue is resolved. Please test out that artifact I linked above and see how it goes. Regarding the overall UWP issue, we're tracking it in the other linked issue above but we're still discussing how to reliably block UWP execution internally.
